src/libraries/System.Text.Json/tests/Serialization/ExceptionTests.cs

```diff
@@ -550,32 +556,38 @@ public static void SerializeUnsupportedType()
         {
             RunTest<Type>(typeof(int));
             RunTest<SerializationInfo>(new SerializationInfo(typeof(Type), new FormatterConverter()));
+            RunTest<IntPtr>((IntPtr)123);
+            RunTest<UIntPtr>((UIntPtr)123);
 
             void RunTest<T>(T value)
             {
-                string fullName = typeof(T).FullName;
+                Type type = typeof(T);
+                string fullName = type.FullName;
 
                 NotSupportedException ex = Assert.Throws<NotSupportedException>(() => JsonSerializer.Serialize(value));
                 string exAsStr = ex.ToString();
                 Assert.Contains(fullName, exAsStr);
                 Assert.Contains("$", exAsStr);
 
-                string serialized = JsonSerializer.Serialize((T)(object)null);
-                Assert.Equal("null", serialized);
-
                 ClassWithType<T> obj = new ClassWithType<T> { Prop = value };
 
                 ex = Assert.Throws<NotSupportedException>(() => JsonSerializer.Serialize(obj));
                 exAsStr = ex.ToString();
                 Assert.Contains(fullName, exAsStr);
                 Assert.Contains("$.Prop", exAsStr);
 
-                obj.Prop = (T)(object)null;
-                serialized = JsonSerializer.Serialize(obj);
-                Assert.Equal(@"{""Prop"":null}", serialized);
+                if (!type.IsValueType)
```

Comment:
Same here, if anything this makes performance worse since `typeof(T).IsValueType` is treated as a JIT-time constant and any branching gets optimized away.